netbox_onboarding/netdev_keeper.py

```diff
@@ -0,0 +1,292 @@
+                message=f"Onboarding for Platform {self.netmiko_device_type} not "
+                f"supported, as it has no specified NAPALM driver",
+            )
+
+    def get_onboarding_facts(
+        self,
+        default_mgmt_if=PLUGIN_SETTINGS["default_management_interface"],
+        default_mgmt_pfxlen=PLUGIN_SETTINGS["default_management_prefix_length"],
+    ):
+        """Gather information from the network device that is needed to onboard the device into the NetBox system.
+
+        Raises:
+          OnboardException('fail-login'):
+            When unable to login to device
+
+          OnboardException('fail-execute'):
+            When unable to run commands to collect device information
+
+          OnboardException('fail-general'):
+            Any other unexpected device comms failure.
+        """
+        # Check to see if the IP address entered was an IP address or a DNS entry, get the IP address
+        self.check_ip()
+
+        self.check_reachability()
+
+        logging.info("COLLECT: device information %s", self.hostname)
+
+        try:
+            # Get Napalm Driver with Netmiko if needed
+            self.set_napalm_driver_name()
+
+            # Raise if no Napalm Driver not selected
+            self.check_napalm_driver_name()
+
+            driver = get_network_driver(self.napalm_driver)
+            optional_args = settings.NAPALM_ARGS.copy()
+            optional_args["secret"] = self.secret
+
+            napalm_device = driver(
+                hostname=self.hostname,
+                username=self.username,
+                password=self.password,
+                timeout=self.timeout,
+                optional_args=optional_args,
+            )
+
+            napalm_device.open()
+
+            logging.info("COLLECT: device facts")
+            self.facts = napalm_device.get_facts()
+
+            logging.info("COLLECT: device interface IPs")
+            self.ip_ifs = napalm_device.get_interfaces_ip()
+
+            try:
+                module_name = PLUGIN_SETTINGS["onboarding_extensions_map"].get(self.napalm_driver)
+                module = importlib.import_module(module_name)
+                driver_addon_class = module.OnboardingDriverExtensions(napalm_device=napalm_device)
+                self.onboarding_class = driver_addon_class.get_onboarding_class()
+                self.driver_addon_result = driver_addon_class.get_ext_result()
+            except ImportError as exc:
+                logging.info(f"No onboarding extension found for driver {self.napalm_driver}")
+
+        except ConnectionException as exc:
+            raise OnboardException(reason="fail-login", message=exc.args[0])
+
+        except CommandErrorException as exc:
+            raise OnboardException(reason="fail-execute", message=exc.args[0])
+
+        except Exception as exc:
+            raise OnboardException(reason="fail-general", message=str(exc))
+
+    def get_netdev_dict(self):
```

Comment:
I wonder if it'd be more maintainable to use a `namedtuple` instead of a `dict` for both `netdev_dict` and `onboarding_kwargs` - that way Python would actually guarantee and enforce that all of the expected keys are present.

That wouldn't work if we want to support the inclusion of additional arbitrary keys though - I can't tell at the moment whether that is a requirement or not?

Reply:
Let's revisit that in the near future